src/zenml/materializers/built_in_materializer.py

```diff
@@ -393,41 +494,157 @@ def save(self, data: Any) -> None:
+
+                # Process elements in chunks to avoid memory issues with very large collections
+                # Use the configurable chunk size from constants
+                chunk_size = DEFAULT_COLLECTION_CHUNK_SIZE
+
+                # Adaptive sizing: if elements are particularly large, reduce chunk size
+                # This is a simple heuristic based on the first few elements
+                if elements and len(elements) > 10:
+                    # Sample the first few elements to estimate size
+                    sample_elements = elements[: min(10, len(elements))]
+                    try:
+                        # Basic size estimation with pickle
+                        avg_size = sum(
+                            len(pickle.dumps(e[1])) for e in sample_elements
+                        ) / len(sample_elements)
+                        # Adjust chunk size based on average element size
+                        # Target max chunk size of ~10MB
+                        target_max_bytes = 10 * 1024 * 1024  # 10MB
+                        if avg_size > 0:
+                            adaptive_chunk_size = max(
+                                1, int(target_max_bytes / avg_size)
+                            )
+                            chunk_size = min(chunk_size, adaptive_chunk_size)
+                    except Exception:
+                        # If size estimation fails, use the default chunk size
+                        pass
+
+                for chunk_idx in range(0, len(elements), chunk_size):
+                    chunk_elements = elements[
+                        chunk_idx : chunk_idx + chunk_size
+                    ]
+                    chunk_indices = [idx for idx, _ in chunk_elements]
+
+                    # Create chunk file path
+                    chunk_filename = f"chunk_{chunk_idx // chunk_size}.pkl.gz"
+                    chunk_path = os.path.join(batch_dir, chunk_filename)
+
+                    # Process objects for serialization
+                    serialized_items = []
+
+                    for _, element in chunk_elements:
+                        # Just validate type compatibility but don't serialize yet
+                        materializer = materializer_class(uri="")
+                        materializer.validate_save_type_compatibility(
+                            element_type
+                        )
+                        serialized_items.append(element)
+
+                    # Write compressed batch to a single file using ZenML fileio
+                    with fileio.open(chunk_path, "wb") as f_raw:
+                        # Use in-memory buffer for gzip compression
+                        with io.BytesIO() as f_buffer:
+                            with gzip.GzipFile(
+                                fileobj=f_buffer, mode="wb", compresslevel=6
+                            ) as f_gzip:
+                                pickle.dump(
```

Comment:
I think this needs a major rework: This code is not using ZenML materializers anymore and instead simply uses pickle for everything. If that was a good solution, instead of doing all this chunking we could just tell people to pickle the whole collection.